src/pygeoprocessing/routing/routing.pyx

```diff
@@ -4273,15 +4282,35 @@ def detect_outlets(d8_flow_dir_raster_path_band, target_outlet_vector_path):
                 if flow_dir == flow_dir_nodata:
                     continue
 
-                # inspect the outflow pixel neighbor
-                flow_dir_n = flow_dir_block[
-                    yi+D8_YOFFSET[flow_dir],
-                    xi+D8_XOFFSET[flow_dir]]
+                is_outlet = 1
+                if d8_flow_dir_mode:
+                    # inspect the outflow pixel neighbor
+                    flow_dir_n = flow_dir_block[
+                        yi+D8_YOFFSET[flow_dir],
+                        xi+D8_XOFFSET[flow_dir]]
+
+                    # if the outflow pixel is outside the raster boundaries or
+                    # is a nodata pixel it must mean xi,yi is an outlet
+                    if flow_dir_n != flow_dir_nodata:
+                        is_outlet = 0
+                else:
+                    # inspect all the outflow pixel neighbors in MFD mode
+                    for n_dir in range(8):
+                        # shift the 0xF mask to the outflow direction and
+                        # test if there's any outflow or not. 0 means nothing
+                        # flows out of that pixel in the ``n_dir`` direction.
+                        if flow_dir&(0xF<<(n_dir*4)) == 0:

```

Comment:
Perhaps this is worth another comment elaborating on how this works? it took me a while to understand:
`0xF<<(n_dir*4)` makes a binary number that has `1111` in the `n_dir`th position of the 32-bit `flow_dir` value. So if all of the 4 flow dir bits are `0`, the bitwise `&` will return `0`.

Question for my own understanding: each of the 8 directions gets its proportion represented in 4 bits. Does that mean we can only represent flow direction in 1/16th = 6.25% increments? Does `1111` mean 100% goes in that direction?

Reply:
Sure, I expanded on that comment from your comment above in 59d8f2715dde0f66789dcc8d97984cb5918897c0

And re your question about max precision, you're close -- the value in the n-th 0xF place is the proportional flow in that direction so you don't know how much really flows in that without knowing how much flows in all other directions. Here's some examples, and assume pixel directions are laid out like this:

321
4x0
567

Then here's some select MFD pixel values and their proportional flow directions
* 0x00000000 -- nodata
* 0x00000001 - 1 proportional unit of flow in the 0 direction with a total sum of 1 means 1/1 -- 100% flows in direction 0
* 0x10000001 - 1 proportional unit of flow in the 0 direction and 1 in in the 7th, so direction 0 gets 1/2 of the flow and direction 7 gets 1/2
* 0xF0000001 - 1 proportional unit of flow in the 0 direction and 15 in in the 7th, so direction 0 gets 1/16 of the flow and direction 7 gets 15/16
* 0xFFFFFFF1 - 1 proportional unit of flow in the 0 direction and 15 in in ALL other directions. So direction 0 gets a proportion of 1/106 (0.94% increment) and all other directions get 15/106 (14.15% increment). This also shows that the smallest increment possible with this configuration is 1/106. 

We thought that wasn't too bad of a precision with also being able to otherwise encode a complicated flow direction in a 32 bit int.

Reply:
Thanks for the explanation! That's really cool that you can store 8 numbers in one.